### PR TITLE
CATTY-226 Implement project migrator

### PR DESCRIPTION
**Pull request by DhrubojyotiBis1**
A `project migrator` service is introduced by the following:

- ProjectMigratorProtocol.swift
- ProjectMigrator.swift

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catty/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Verify that the Jira ticket is in the status *Ready for Development*
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *#catty* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Comment by m-herold** [comment]
And I am curious: how did you include those translations? Did you fetch them from Crowdin?

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        guard let versionNumber = Float(project.header.catrobatLanguageVersion) else {
+            throw ProjectMigratorError.unknown(description: "Unable to convert version number to Float")
+        }
+
+        if versionNumber < self.correctVersionNumber {
+            throw ProjectMigratorError.invalidCatrobatLanguageVersion
+        }
+
+        self.projectPath = project.projectPath()
+        let fileManager = CBFileManager()
```
Comment (m-herold) [inline src/Catty/Parser&Serialializer/ProjectMigrator.swift:48]: Do not use this dependency like that directly in the middle of your production code. This makes this function untestable. Use constructor dependency injection and assign this dependency to a member variable. That enables you to replace the production version with a test double while unit testing this code.

**Comment by DhrubojyotiBis1** [comment]
Hi @m-herold I have made the changes although I am not sure if my approach of creating a Project which is passed into the `migrateToScene(project:)`  function is correct. Please review

**Comment by DhrubojyotiBis1** [comment]
> And I am curious: how did you include those translations? Did you fetch them from Crowdin?

I used google translate 😛 
Is there any specific API from Crowdin which is supposed to be used?

**Comment by m-herold** [comment]
Thanks for the effort :) However, this is how this process is automated in our project:

- In general translations are handled via Crowdin where they get manually translated by a large number of native speaker volunteers
- Our source strings (strings to be translated) are uploaded to the Crowdin platform every night at 00:00:00 (https://github.com/Catrobat/Catty/blob/develop/.github/workflows/scheduled-upload-crowdin.yml)
- The upload and download of translations to and from Crowdin is all automated using a _fastlane_ lane (i.e., https://github.com/Catrobat/Catty/blob/develop/src/fastlane/Fastfile#L143)
- This lane is also called in the release lane, meaning that this process is completely automated in our release Jenkins pipeline (https://github.com/Catrobat/Catty/blob/develop/Jenkinsfile.release#L38)

In other words, that means that although you made a lot of effort by "manually" translating these strings, they will all be overwritten. So that is how new translations should be added: 

- add them to the translation define file, just like you did
- when you run Catty after that, the new source strings will be copied to the default translation (English)
- everything else happens automatically

For your pull request, please remove any modifications to translation files other than the default translation.

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+class ProjectMigrator: ProjectMigratorProtocol {
+
+    private let correctVersionNumber: Float = 0.992
+    private var projectPath: String = ""
```
Comment (m-herold) [inline src/Catty/Parser&Serialializer/ProjectMigrator.swift:26]: What are the _projectPath_ and _directoryPath_ do? Is there really a need to save state in member variables?

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    private var directoryPath: String = ""
+    var fileManager: CBFileManager!
+
+    enum FileSystem: String {
```
Comment (m-herold) [inline src/Catty/Parser&Serialializer/ProjectMigrator.swift:30]: I would suggest renaming it to _ProjectResouce_ which I think would be more self-explaining.

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                throw error
+            }
+
+        } else {
+            throw ProjectMigratorError.unknown(description: "Unable to create \(kLocalizedScene) directory")
+        }
+
+    }
+
+    private func changeDirectory(for fileOrDirectory: String, isFile: Bool, isNecessary: Bool) throws {
```
Comment (m-herold) [inline src/Catty/Parser&Serialializer/ProjectMigrator.swift:78]: What does this method do?

Comment (DhrubojyotiBis1) [reply src/Catty/Parser&Serialializer/ProjectMigrator.swift]: Hi @m-herold  I have changed the name of the function to `migrateToSceneDirectory(for projectResouce:, isFile:, shouldThrowIfRequired:)` to make itself explanatory.  This function moves each file/directory to the new directory `/Scene`.

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift (hunk 2 above)
Comment (DhrubojyotiBis1) [reply src/Catty/Parser&Serialializer/ProjectMigrator.swift:26]: Hi @m-herold  `directoryPath` is not required so I removed it but `projectPath` is required as it is being used in the `migrateToSceneDirectory(for projectResouce:, isFile:, shouldThrowIfRequired:)` function.

please correct me if I am doing something wrong.

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+class ProjectMigrator: ProjectMigratorProtocol {
+
+    private let correctVersionNumber: Float = 0.992
```
Comment (m-herold) [inline src/Catty/Parser&Serialializer/ProjectMigrator.swift:25]: Please rename to `minimumCatrobatLanguageVersionForScenes` and make it public (can be reused in the tests).

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    private var projectPath: String = ""
+    var fileManager: CBFileManager!
```
Comment (m-herold) [inline src/Catty/Parser&Serialializer/ProjectMigrator.swift:27]: The implicit force unwrap is not necessary here, since the `fileManager` is passed in the initializer and is non-optional.

File: src/CattyTests/Project/ProjectMigratorTest.swift [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+        self.header = Header()
+        header.catrobatLanguageVersion = "0.992"
+
+        project = Project()
+        project.header = header
+    }
+
+    func testMigrateToScene() {
+        var automaticScreenShortAtPath = "\(String(describing: project.projectPath()))automatic_screenshot.png"
```
Comment (m-herold) [inline src/CattyTests/Project/ProjectMigratorTest.swift:45]: Please correct the typo.

File: src/CattyTests/Project/ProjectMigratorTest.swift [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    var newFolderName = String()
+    var header: Header!
+    var project: Project!
+
+    override func setUp() {
+        super.setUp()
+        newFolderName = kLocalizedScene
+
+        self.header = Header()
+        header.catrobatLanguageVersion = "0.992"
```
Comment (m-herold) [inline src/CattyTests/Project/ProjectMigratorTest.swift:38]: Make the constant from the migrator public and use directly here.

File: src/CattyTests/Project/ProjectMigratorTest.swift [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+import XCTest
+
+@testable import Pocket_Code
+
+class ProjectMigratorTest: XCTestCase {
+
+    var newFolderName = String()
```
Comment (m-herold) [inline src/CattyTests/Project/ProjectMigratorTest.swift:29]: Please use a persistent code style and remove the initializer here (like you did for `project` and `header`).

File: src/CattyTests/Project/ProjectMigratorTest.swift [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            try migrate.migrateToScene(project: project)
+        } catch {
+            XCTFail("Some error occured")
+        }
+
+        automaticScreenShortAtPath = "\(String(describing: project.projectPath()) + newFolderName) 1/automatic_screenshot.png"
+        imageDirectoryAtPath = "\(String(describing: project.projectPath()) + newFolderName) 1/images"
+        soundDirectoryAtPath = "\(String(describing: project.projectPath() + newFolderName)) 1/sounds"
+
+        XCTAssertTrue(fileManager.fileExists(automaticScreenShortAtPath))
```
Comment (m-herold) [inline src/CattyTests/Project/ProjectMigratorTest.swift:63]: Please also verify that the other, old files do not exist anymore, since the migrator should move those files and not create new files/directories.

File: src/Catty/Defines/LanguageTranslationDefines.h [hunk 11]
```diff
@@ -279,6 +280,7 @@
 #define kLocalizedRemovedKnownBluetoothDevices NSLocalizedString(@"All known Bluetooth devices successfully removed", nil)
 #define kLocalizedArduinoBricksDescription NSLocalizedString(@"Allow the app to control Arduino boards", nil)
 
+
```
Comment (m-herold) [inline src/Catty/Defines/LanguageTranslationDefines.h:283]: Please remove newline.

File: src/CattyTests/Project/ProjectMigratorTest.swift [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        let imageDirectoryAtPath = "\(String(describing: project.projectPath()))images"
+        let soundDirectoryAtPath = "\(String(describing: project.projectPath()))sounds"
+
+        let fileManager = CBFileManagerMock(filePath: [] as! [String], directoryPath: [imageDirectoryAtPath, soundDirectoryAtPath])
+
+        let migrate = ProjectMigrator(fileManager: fileManager)
+
+        do {
+
+            try migrate.migrateToScene(project: project)
```
Comment (m-herold) [inline src/CattyTests/Project/ProjectMigratorTest.swift:97]: Please use the Nimble framework to do that assertion (https://github.com/Quick/Nimble#swift-error-handling). This framework is already included and used in our project.

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                throw error
+            }
+
+        } else {
+            throw ProjectMigratorError.unknown(description: "Unable to create \(kLocalizedScene) directory")
+        }
+
+    }
+
+    private func migrateToSceneDirectory(for projectResouce: ProjectResouce, isFile: Bool, shouldThrowIfRequired: Bool) throws {
```
[the remark this answers is not in the document]
Comment (DhrubojyotiBis1) [reply src/Catty/Parser&Serialializer/ProjectMigrator.swift:77]: I am using this `isFile` argument to check whether its a directory or a file that is to be moved to a new `/Scene` directory since it is required at the time of moving the file/directory which uses `moveExistingFile` and ` moveExistingDirectory` function of the `CBFileManager` respectively.

Please correct me if there is a better alternative.

**Comment by DhrubojyotiBis1** [comment]
Hi, @m-herold I have added tests documenting `CBFileManagerMock`. Please review.

**Comment by m-herold** [comment]
Thanks. The new test looks very good!

File: src/CattyTests/Project/ProjectMigratorTest.swift [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        let projectPath = project.projectPath()
+
+        let imageDirectoryAtPath = "\(String(describing: project.projectPath()))images"
+        let soundDirectoryAtPath = "\(String(describing: project.projectPath()))sounds"
+
+        let fileManager = CBFileManagerMock(filePath: [] as! [String], directoryPath: [imageDirectoryAtPath, soundDirectoryAtPath, projectPath])
+
+        let migrate = ProjectMigrator(fileManager: fileManager)
+
+        expect { try migrate.migrateToScene(project: self.project) }.to(throwError(ProjectMigratorError.filePathNotFound))
```
Comment (m-herold) [inline src/CattyTests/Project/ProjectMigratorTest.swift:112]: Which filePath is not found in this test? In general the behavior should be as follows:

There are certain projects which do not contain an _images_ or _sounds_ folder or any screenshot files. However, these projects should also be migrated, so there should not be any error in that case. Please change the test and the implementation in order to do so.

PS: ideally, in case there is no _images_ or _sounds_ directory available, please create it in the scene folder. Thanks!

Comment (DhrubojyotiBis1) [reply src/CattyTests/Project/ProjectMigratorTest.swift:112]: @m-herold  so the function `migrateToScene(project:)` should only throw an error if `automatic_screenshot.png` file is not present and all other files/directory are optional.

should this be the approach ?

Comment (m-herold) [reply src/CattyTests/Project/ProjectMigratorTest.swift:112]: When a project has not been started yet, then this file does not exist either (no error in this case please).

**Comment by DhrubojyotiBis1** [comment]
> I have now had a closer look at the actual implementation and left some comments. In general, please make use of `kProjectImagesDirName`, `kProjectSoundsDirName`, `kProjectCodeFileName`, `kScreenshotFilename`, `kScreenshotManualFilename` and `kScreenshotAutoFilename` instead of using those strings directly.

Hi, @m-herold I have made the changes. Please review

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+    }
+
+    private func moveToSceneDirectory(for projectResouceName: String, projectPath: String, isFile: Bool) {
+
+        let fileOrDirectoryAtPath = projectPath + projectResouceName
+        let fileOrDirectoryToPath = "\(projectPath + kLocalizedScene) 1/\(projectResouceName)"
+
+        if isFile {
```
Comment (m-herold) [inline src/Catty/Parser&Serialializer/ProjectMigrator.swift:71]: Please change to `isFile and fileManager.fileExists` to reduce cyclomatic complexity.

File: src/Catty/Parser&Serialializer/ProjectMigrator.swift [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+            if !fileManager.fileExists(fileOrDirectoryAtPath) {
+                return
+            }
+
+            fileManager.moveExistingFile(atPath: fileOrDirectoryAtPath, toPath: fileOrDirectoryToPath, overwrite: false)
+        } else {
```
Comment (m-herold) [inline src/Catty/Parser&Serialializer/ProjectMigrator.swift:78]: Please change to`else if fileManager.directoryExists` to reduce cyclomatic complexity.

**Comment by DhrubojyotiBis1** [comment]
Hi @m-herold  I have made the changes. Please review

**Comment by m-herold** [comment]
Very nice pull request! Thanks!
